### PR TITLE
[validator set] adding basic validator set generation

### PR DESCRIPTION
**Pull request by sblackshear**
Implement a starting point for storing the validator set on-chain with an eye toward supporting reconfiguration in the future. This modifies the ValidatorSet set module to add:

- A ValidatorConfig resource that candidate validators can publish under their accounts. This resource holds all of the public keys of the validator and can be used to support key rotation. Anyone can publish such a resource under their account; it doesn't convey any special privileges.
- A procedure that can add an address to the validator set by grabbing the ValidatorConfig resource from that address and adding the keys to the ValidatorSet.T vector. This procedure only works partially because the Vector.push_back native function has not yet been implemented.

In addition, this modifies the genesis transaction generation code to add the initial validator set.

## Test Plan

New end-to-end tests. Will add more once `Vector.push_back` works.

### CODE REVIEW

File: language/stdlib/modules/validator_set.mvir [hunk 1]
```diff
@@ -1,30 +1,101 @@
...
+
+        config_ref = borrow_global_mut<ValidatorConfig>(get_txn_sender());
+        keys_ref = &mut move(config_ref).keys;
+        *(&mut move(keys_ref).consensus_pubkey) = move(consensus_pubkey);
+
+        return;
+    }
+
+    // TODO: Decide on access control policy. For now, we ensure that this is only callable from the
+    // genesis txn. Obviously, we'll need a different policy once we support reconfiguration.
```
Comment (aching) [inline language/stdlib/modules/validator_set.mvir]: I remember one of our ideas were that a caller could that scans account addresses for LITs and ValidatorConfig and then just set the validators with that config if it changed?  I'm not sure how this method fits in.

Comment (sblackshear) [reply language/stdlib/modules/validator_set.mvir]: Good question. This fits into that model. Basically, the Association (or whoever has the authority to set the validator set) would write a transaction script with a list of a validator addresses and call `add_validator` for each address in the list. This `add_validator` code is what checks the LITs and ValidatorConfig.

File: language/stdlib/modules/validator_set.mvir [hunk 2]
```diff
@@ -1,30 +1,93 @@
...
+    struct ValidatorInfo {
+        addr: address,
+        network_signing_pubkey: bytearray,
+        network_identity_pubkey: bytearray,
+        consensus_pubkey: bytearray,
+        // TODO: voting power
+    }
+
+    // A current or prospective validator should publish one of these under their address
+    resource ValidatorConfig {
```
Comment (zekun000) [inline language/stdlib/modules/validator_set.mvir]: what's the difference between the `Config` and `Info` and `resource` vs `struct`?

File: language/stdlib/modules/validator_set.mvir [hunk 3]
```diff
@@ -1,30 +1,93 @@
...
+            });
+
+        return;
     }
+
+    // Get the consensus public key for the validator candidate at the given address
+    public get_consensus_pubkey(addr: address): bytearray acquires ValidatorConfig {
+        let config_ref: &Self.ValidatorConfig;
+
+        config_ref = borrow_global<ValidatorConfig>(move(addr));
```
Comment (zekun000) [inline language/stdlib/modules/validator_set.mvir]: `borrow_global` means borrowing the <resource> from the address?

File: language/stdlib/modules/validator_set.mvir [hunk 4]
```diff
@@ -1,30 +1,93 @@
...
+        return *&move(config_ref).consensus_pubkey;
+    }
+
+    // Rotate a validator candidate's consensus public key. The change will not take effect until
+    // the next reconfiguration.
+    public rotate_consensus_pubkey(consensus_pubkey: bytearray) acquires ValidatorConfig {
+        let config_ref: &mut Self.ValidatorConfig;
```
Comment (zekun000) [inline language/stdlib/modules/validator_set.mvir]: it's a little confusing to me which functions apply to each validator and which apply to the ValidatorSet

**Pull request merged**